src/baseTrie.ts

```diff
@@ -122,278 +114,210 @@ export class Trie {
    * @method get
    * @memberof Trie
    * @param {Buffer} key - the key to search for
-   * @param {Function} cb A callback `Function` which is given the arguments `err` - for errors that may have occured and `value` - the found value in a `Buffer` or if no value was found `null`
+   * @returns {Promise} - Returns a promise that resolves to `Buffer` if a value was found or `null` if no value was found.
    */
-  get(key: Buffer, cb: BufferCallback) {
-    key = ethUtil.toBuffer(key)
-
-    this.findPath(key, (err: Error, node: TrieNode, remainder: number[], stack: TrieNode[]) => {
-      let value = null
-
-      if (node && remainder.length === 0) {
-        value = node.value
-      }
-
-      cb(err, value)
-    })
+  async get(key: Buffer): Promise<Buffer | null> {
+    const { node, remaining } = await this.findPath(key)
+    let value = null
+    if (node && remaining.length === 0) {
+      value = node.value
+    }
+    return value
   }
 
   /**
    * Stores a given `value` at the given `key`
    * @method put
    * @memberof Trie
    * @param {Buffer} key
-   * @param {Buffer} Value
-   * @param {Function} cb A callback `Function` which is given the argument `err` - for errors that may have occured
+   * @param {Buffer} value
+   * @returns {Promise}
    */
-  put(key: Buffer, value: Buffer, cb: ErrorCallback) {
-    key = ethUtil.toBuffer(key)
-    value = ethUtil.toBuffer(value)
-
+  async put(key: Buffer, value: Buffer): Promise<void> {
+    // If value is empty, delete
     if (!value || value.toString() === '') {
-      this.del(key, cb)
+      return await this.del(key)
+    }
+
+    await this.lock.wait()
+    if (this.root.equals(KECCAK256_RLP)) {
+      // If no root, initialize this trie
+      await this._createInitialNode(key, value)
     } else {
-      cb = callTogether(cb, this.sem.leave)
-
-      this.sem.take(() => {
-        if (this.root.toString('hex') !== ethUtil.KECCAK256_RLP.toString('hex')) {
-          // first try to find the give key or its nearst node
-          this.findPath(
-            key,
-            (err: Error, foundValue: TrieNode, keyRemainder: number[], stack: TrieNode[]) => {
-              if (err) {
-                return cb(err)
-              }
-              // then update
-              this._updateNode(key, value, keyRemainder, stack, cb)
-            },
-          )
-        } else {
-          this._createInitialNode(key, value, cb) // if no root initialize this trie
-        }
-      })
+      // First try to find the given key or its nearest node
+      const { remaining, stack } = await this.findPath(key)
+      // then update
+      await this._updateNode(key, value, remaining, stack)
     }
+    this.lock.signal()
```

Comment:
This new async/await compatible locking is really great! Generally functions like the above (`put`) are so much more readable now, phew! 👍

Also thanks for all the small improvements like changing execution order on conditionals, rethinking the appliance of asserts, code cleanups,... 🎊